src/lib/traceCall.js

```diff
@@ -0,0 +1,199 @@
+// @flow
+import { EthRpc, convert } from 'emerald-js';
+import BigNumber from 'bignumber.js';
+
+type Call = {
+    method: string,
+    params: Array<any>,
+}
+
+type Transaction = {
+    from: string,
+    to: string,
+    gas: string,
+    gasPrice: string,
+    value: string,
+    data: string,
+}
+
+interface ITracer {
+    buildRequest(): Call;
+    estimateGas(trace: any): ?BigNumber;
+}
+
+/**
+ * Every Ethereum implementation has eth_estimateGas
+ */
+class CommonCallTracer implements ITracer {
+    tx: Transaction;
+
+    constructor(tx: Transaction) {
+        this.tx = tx;
+    }
+
+    buildRequest(): Call {
+        const params = [{
+            from: this.tx.from,
+            to: this.tx.to,
+            gas: this.tx.gas,
+            gasPrice: this.tx.gasPrice,
+            value: this.tx.value,
+            data: this.tx.data,
+        }];
+        params.push('latest');
+
+        return {
+            method: 'eth_estimateGas',
+            params,
+        };
+    }
+
+    estimateGas(trace: any): BigNumber | null {
+        return convert.toBigNumber(trace);
+    }
+}
+
+/**
+ * Parity has trace_call API method
+ */
+export class ParityTracer implements ITracer {
+    tx: Transaction;
+
+    constructor(tx: Transaction) {
+        if (!tx) {
+            throw new Error(`Invalid value of tx ${tx}`);
+        }
+        this.tx = tx;
+    }
+
+    buildRequest(): Call {
+        const params = [{
+            from: this.tx.from,
+            to: this.tx.to,
+            gas: this.tx.gas,
+            gasPrice: this.tx.gasPrice,
+            value: this.tx.value,
+            data: this.tx.data,
+        }];
+        // Should we use 'vmTrace' or not ?
```

Comment:
it is not necessary if you are using 'stateDiff', but it  does not hurt to add it either.